cpp/src/io/parquet/compact_protocol_reader.cpp

```diff
@@ -40,37 +42,57 @@ class parquet_field {
   int field() const { return _field_val; }
 };
 
+std::string field_type_string(FieldType type)
+{
+  switch (type) {
+    case ST_FLD_TRUE: return "bool(true)";
+    case ST_FLD_FALSE: return "bool(false)";
+    case ST_FLD_BYTE: return "int8";
+    case ST_FLD_I16: return "int16";
+    case ST_FLD_I32: return "int32";
+    case ST_FLD_I64: return "int64";
+    case ST_FLD_DOUBLE: return "double";
+    case ST_FLD_BINARY: return "binary";
+    case ST_FLD_STRUCT: return "struct";
+    case ST_FLD_LIST: return "list";
+    case ST_FLD_SET: return "set";
+    default: return "unknown(" + std::to_string(type) + ")";
+  }
+}
+
+void assert_field_type(int type, FieldType expected)
+{
+  CUDF_EXPECTS(type == expected,
+               "expected " + field_type_string(expected) + " field, got " +
+                 field_type_string(static_cast<FieldType>(200)) + " field instead");
+}
+
 /**
  * @brief Abstract base class for list functors.
  */
-template <typename T>
+template <typename T, FieldType EXPECTED_ELEM_TYPE>
 class parquet_field_list : public parquet_field {
  private:
-  using read_func_type = std::function<bool(uint32_t, CompactProtocolReader*)>;
-  FieldType _expected_type;
+  using read_func_type = std::function<void(uint32_t, CompactProtocolReader*)>;
   read_func_type _read_value;
 
  protected:
   std::vector<T>& val;
 
   void bind_read_func(read_func_type fn) { _read_value = fn; }
 
-  parquet_field_list(int f, std::vector<T>& v, FieldType t)
-    : parquet_field(f), _expected_type(t), val(v)
-  {
-  }
+  parquet_field_list(int f, std::vector<T>& v) : parquet_field(f), val(v) {}
 
  public:
-  inline bool operator()(CompactProtocolReader* cpr, int field_type)
+  inline void operator()(CompactProtocolReader* cpr, int field_type)
   {
-    if (field_type != ST_FLD_LIST) { return true; }
+    assert_field_type(field_type, ST_FLD_LIST);
```

Comment:
Probably we are wrong from the beginning. Why not `field_type` an enum value, instead of `int`?

Reply:
In addition, should the caller handle the input type instead of checking it here?

Reply:
`field_type` is obtained by the caller by parsing the encoded input, so it could be any value if the input file is garbage. We cannot assume it will be a valid FieldType.

As for the validation on the caller side - the caller only sees these parquet_field_xyz objects as functors and does not know the actual field type that is being read. We could move validation to the caller if `parquet_field` classes somehow expose the expected field type, but I don't think this is a better option.